### PR TITLE
The qpack encoder should write all instructions directly into a transport stream

### PR DESCRIPTION
**Pull request by ddragana**
Probably this is all that is needed for #655

### CODE REVIEW

File: neqo-qpack/src/encoder.rs [hunk 1]
```diff
@@ -257,28 +271,42 @@ impl QPackEncoder {
...
+
+    fn maybe_send_change_capacity(&mut self, conn: &mut Connection, stream_id: u64) -> Res<()> {
+        if let Some(cap) = self.change_capacity {
+            let mut buf = QPData::default();
+            EncoderInstruction::Capacity { value: cap }.marshal(&mut buf, self.use_huffman);
+            if !conn.stream_send_atomic(stream_id, &buf)? {
+                return Err(Error::EncoderStreamBlocked);
+            }
+            self.table.set_capacity(cap)?;
+            self.change_capacity = None;
```
Comment (agrover) [inline neqo-qpack/src/encoder.rs:287]: If you use `self.change_capacity.take()` ([ref](https://doc.rust-lang.org/std/option/enum.Option.html#method.take)) you don't need to assign to `None` at the bottom.

File: neqo-qpack/src/encoder.rs [hunk 2]
```diff
@@ -81,7 +97,8 @@ impl QPackEncoder {
         let new_cap = std::cmp::min(self.max_table_size, cap);
         self.max_entries = new_cap / 32;
         // we also set our table to the max allowed.
-        self.change_capacity(new_cap)
+        self.change_capacity(new_cap);
```
Comment (agrover) [inline neqo-qpack/src/encoder.rs:100]: So the max capacity might not actually change? Maybe we don't want to call `change_capacity()` if it hasn't actually increased?

File: neqo-qpack/src/encoder.rs [hunk 3]
```diff
@@ -257,28 +271,42 @@ impl QPackEncoder {
...
-            let r = conn
-                .stream_send(stream_id, &self.send_buf[..])
-                .map_err(|_| Error::EncoderStream)?;
-            qdebug!([self], "{} bytes sent.", r);
-            self.send_buf.read(r as usize);
-            Ok(())
-        } else {
-            Ok(())
+        match self.local_stream {
+            LocalStreamState::NoStream => Ok(()),
```
Comment (agrover) [inline neqo-qpack/src/encoder.rs:297]: Just wanted to check this should be `Ok(())`. Is calling `send()` allowed when there isn't even an uninitialized stream? Does this deserve to be an error or emit a log message?

File: neqo-qpack/src/encoder.rs [hunk 4]
```diff
@@ -509,6 +537,12 @@ mod tests {
         })
     }
 
+    fn change_capacity(encoder: &mut TestEncoder, capacity: u64) -> Res<()> {
```
Comment (martinthomson) [inline neqo-qpack/src/encoder.rs:540]: Can you make this a method on TestEncoder?

File: neqo-qpack/src/encoder.rs [hunk 5]
```diff
@@ -257,28 +271,44 @@ impl QPackEncoder {
...
-    fn change_capacity(&mut self, value: u64) -> Res<()> {
+    fn need_change_capacity(&mut self, value: u64) {
         qdebug!([self], "change capacity: {}", value);
-        self.table.set_capacity(value)?;
-        EncoderInstruction::Capacity { value }.marshal(&mut self.send_buf, self.use_huffman);
+        self.change_capacity = Some(value);
+    }
+
+    fn maybe_send_change_capacity(&mut self, conn: &mut Connection, stream_id: u64) -> Res<()> {
+        if let Some(cap) = self.change_capacity.take() {
```
Comment (martinthomson) [inline neqo-qpack/src/encoder.rs:280]: If this write fails, then you won't send the capacity change.  So you should defer clearing the value until the write succeeds.

File: neqo-qpack/src/encoder.rs [hunk 6]
```diff
@@ -257,28 +271,44 @@ impl QPackEncoder {
...
+            let mut buf = QPData::default();
+            EncoderInstruction::Capacity { value: cap }.marshal(&mut buf, self.use_huffman);
+            if !conn.stream_send_atomic(stream_id, &buf)? {
+                return Err(Error::EncoderStreamBlocked);
+            }
+            self.table.set_capacity(cap)?;
```
Comment (martinthomson) [inline neqo-qpack/src/encoder.rs:286]: As we have two fallible operations here, this is tricky.  I would suggest that you try to this (and attempt to perform evictions) first as that operation - if it succeeds - won't result in a half-committed state.  If we manage to set capacity, but not send the instruction, then we can still try to set the capacity again later before writing another instruction.  If we can't evict, but send the instruction, then we will have violated the protocol (probably).

File: neqo-qpack/src/encoder.rs [hunk 7]
```diff
@@ -257,28 +271,44 @@ impl QPackEncoder {
         }
     }
 
-    fn change_capacity(&mut self, value: u64) -> Res<()> {
+    fn need_change_capacity(&mut self, value: u64) {
```
Comment (martinthomson) [inline neqo-qpack/src/encoder.rs:274]: Names are hard, but I think that a function called `change_capacity` is very clear.  If anything, the variable name is what needs changing, to `next_capacity` or something perhaps.

File: neqo-qpack/src/encoder.rs [hunk 8]
```diff
@@ -14,21 +14,36 @@ use crate::stats::Stats;
...
 use std::collections::{HashMap, HashSet, VecDeque};
 use std::convert::TryFrom;
 
 pub const QPACK_UNI_STREAM_TYPE_ENCODER: u64 = 0x2;
 
+#[derive(Debug, PartialEq)]
+enum LocalStreamState {
+    NoStream,
+    Uninitialized(u64),
+    Initialized(u64),
```
Comment (martinthomson) [inline neqo-qpack/src/encoder.rs:28]: If these are stream IDs, can you use `StreamId`.  If for no other reason than we have far too many u64s floating around.

File: neqo-qpack/src/encoder.rs [hunk 9]
```diff
@@ -264,28 +278,56 @@ impl QPackEncoder {
...
+    fn change_capacity(&mut self, value: u64) {
         qdebug!([self], "change capacity: {}", value);
-        self.table.set_capacity(value)?;
-        EncoderInstruction::Capacity { value }.marshal(&mut self.send_buf, self.use_huffman);
+        self.next_capacity = Some(value);
+    }
+
+    fn maybe_send_change_capacity(&mut self, conn: &mut Connection, stream_id: u64) -> Res<()> {
+        if let Some(cap) = self.next_capacity {
+            // Check if itt is possible to reduce the capacity, e.g. if enough space can be make free for the reduction.
```
Comment (martinthomson) [inline neqo-qpack/src/encoder.rs:288]:
```suggestion
            // Check if it is possible to reduce the capacity, e.g. if enough space can be make free for the reduction.
```

File: neqo-qpack/src/encoder.rs [hunk 10]
```diff
@@ -264,28 +278,56 @@ impl QPackEncoder {
...
+            if cap < self.table.capacity() && !self.table.test_evict_to(cap) {
```
Comment (martinthomson) [inline neqo-qpack/src/encoder.rs:289]: This might be better named `can_evict_to`

File: neqo-qpack/src/encoder.rs [hunk 11]
```diff
@@ -264,28 +278,56 @@ impl QPackEncoder {
...
+            }
+            let mut buf = QPData::default();
+            EncoderInstruction::Capacity { value: cap }.marshal(&mut buf, self.use_huffman);
+            if !conn.stream_send_atomic(stream_id, &buf)? {
+                return Err(Error::EncoderStreamBlocked);
+            }
+            if self.table.set_capacity(cap).is_err() {
+                debug_assert!(
+                    false,
+                    "test_evict_to should have checked and make sure this operation is possible"
```
Comment (martinthomson) [inline neqo-qpack/src/encoder.rs:300]: If you change the name:
```suggestion
                    "can_evict_to should have checked and make sure this operation is possible"
```

File: neqo-qpack/src/encoder.rs (hunk 8 above)
Comment (ddragana) [reply neqo-qpack/src/encoder.rs:28]: I will switch it to StreamId here.

**Pull request merged**